### DISCUSSION TITLE
tree shaking에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-woowahan-bansanghwe#75] [Optimization]
## Tree shaking이란 무엇인가요?

**사용되지 않는 코드를 제거**하기 위해 JS 컨텍스트에서 일반적으로 사용되는 용어이다. 이 용어는 rollup에 의해 인기를 얻게 되었고, webpack에서도 찾아볼 수 있는 개념이다. 

나무를 흔들어서 죽은 나뭇잎들을 떨어뜨리듯, 코드를 빌드할 때 쓰이지 않는 코드를 제외한다는 뜻으로 tree shaking이라는 이름이 붙여졌다. 

어플리케이션이 복잡해지면서 많은 dependency가 추가된다. 개발자가 직접 사용되지 않는 모듈을 제거하는 것은 많은 비용이 든다. 

tree shaking은 정적 ES6 모듈의 특정 부분을 가져오는 import 구문의 이점을 사용해 이러한 문제를 해결한다. 

## 문제

- 자바스크립트 파일의 크기가 점점 커지고 있다. 이는 파서와 컴파일러에게 부담을 주고, 처리하는 데에 많은 비용을 소모하게 한다. 다른 리소스(이미지 등)에 비해 비싼 리소스라고 할 수 있다. 이때 사용되지 않는 코드를 파일에서 제거할 수 있다면 엄청난 최적화가 되지 않을까? 이 문제를 해결하기 위해 등장한 것이 tree shaking이다. 
- tree shaking이 복잡한 이유는 또 있다. 개발자가 작성하고 있는 코드라면 사용되지 않는 코드. 즉, import 되지 않는 모듈을 삭제하는 일은 쉬울 수 있다. 하지만 라이브러리라면 이야기가 달라진다. A라이브러리에서 a, b모듈을 사용하고 있고, c, d 등의 모듈을 사용하고 있지 않을 때 c, d 등의 모듈을 삭제하는 일은 상당히 번거롭다. 어떤 side effect가 있을 지 잘 알 수 없기 때문이다.

### 세세한 문제 (wild card import)

```jsx
import * as utils from 'utils';
// utils 모듈로부터 모든 것을 import해서 utils 네임스페이스에 넣어라. 
// 하지만 우리는 보통 utils 네임스페이스 안 모든 것을 사용하지는 않는다.
```

```jsx
import { simpleSort } from 'utils';
// utils 모듈에서 simpleSort만 가져와! 
// 이는 트리 쉐이킹을 위한 작업이라고 볼 수 있다. 
```

utils에는 많은 모듈 코드가 있을 것이고 우리는 utils에 포함된 모든 모듈을 사용하진 않는다. 와일드카드로 import한 utils를 그대로 빌드한다면 utils에 포함된 모든 코드가 포함되어 빌드가 될 것이다. 

> wild card import을 대부분 모듈의 index 파일을 작성할 때 사용하는데 (import와 동시에 export 해주고 싶어서) 생각해보면 tree shaking이 안될 것 같은데 된다?!..!
>  
> 이는 webpack4에 생긴 providedExports 최적화 옵션을 보면된다. (export * from 을 위해 만들어진 옵션..! 기본값이 true이므로 따로 설정하지 않아도 re-export 구문을 최적화한다. - [참고](https://webpack.js.org/configuration/optimization/#optimizationprovidedexports)
> 

## code splitting vs tree shaking

코드 스플릿팅은 자바스크립트 **청크**로 어플리케이션을 분할하고, 그 청크를 **필요로 하는 경로에만 이 청크들을 배분하여 성능을 개선하는 기술**이다. 이 기술을 사용한다고 해서 사용되지 않는 코드를 포함한 무거운 자바스크립트 어플리케이션의 일반적인 문제를 해결하지는 못한다. 이 문제의 해법을 tree shaking에서 찾을 수 있다. 

## 정적 import

es2015의 모듈 구문은 정적구조에 의존한다. 이에 두가지 제약이 있는데 

첫째는, import 문에 동적 매개변수를 사용할 수 없다는 것

둘째, 런타임이나 조건부로 모듈을 불러올 수 없다는 것이다. 

이유는 import와 export는 코드 구조의 중심을 잡아주는 역할을 하기 때문이다. 코드 구조를 분석해 모듈을 한데 모아 번들링하고, 사용하지 않는 모듈은 제거 (tree shaking)해야 하는 데, 이는 코드 구조가 간단하고 고정되어 있을 때만 가능한 작업이기 때문이다. 

이런 이유로 tree shaking은 commonJS에서 하기 어렵다. (require는 동적으로 모듈을 가져올 수 있다. 즉 조건부로 모듈을 가져올 수 있는데 이는 런타임이 아닌 경우 어떤 모듈을 가져오는지 아닌지 알 수 없다!) 

## webpack에서의 tree shaking

webpack은 es module로 의존성을 관리하는 모듈만 tree shaking을 한다. 

webpack2 이후 사용하지 않는 모듈의 export를 감지하는 기능을 제공한다. 

따라서, require로 변환된 코드에서는 트리 쉐이킹이 적용되지 않는다. 

### side effect

또 webpack4 이후 sideEffect를 통해 웹팩에게 힌트를 주어 사용하지 않으므로 제거해도 된다고 표시한다. (웹팩에게 힌트를 주어 실수로 삭제되는 것을 방지한다.) 

어플리케이션에서 사용하지 않는 디펜던시를 제거할 때 고려해야할 점은 프로젝트의 모듈이 사이드 이펙트를 발생시키는지에 대한 여부이다. 예측 가능한 입력을 가지고 동일하게 함수의 스코프 밖의 어떤 것도 변경하지 않으면서 예측 가능한 결과를 반환하는 모듈이 안전하게 트리 쉐이킹을 할 수 있는 디펜던시이다.

즉, 현재 모듈 외에 다른 모듈에 영향을 끼치는 요소가 있다면 이를 side effect가 있다고 할 수 있다. 

```jsx
// 예시 

let animals = ['dog', 'cat'];

const addAnimals = (name) => {
  animals.push(name);
}

/*
만약 addAnimals() 함수가 쓰이지 않고 있다고 해도,
addAnimals()가 함수 바깥 변수를 변경하기 때문에 
side effect를 일으킨다고 판단하여 
트리쉐이킹을 하지 못한다. 
*/
```

이런 모듈들이 있을 수 있는데 해당 개발자가 봤을 때 이는 side effect를 발생시키지 않는다! 라고 판단할 경우 side effect가 일어나지 않는 모듈이라고 설정할 수 있다. 

웹팩은 기본적으로 모듈에 side effect가 있다고 전제하기 때문에 side effect가 없다는 것을 명시해주지 않으면 빌드시 코드를 삭제하지 않고 포함시킨다. 즉 side effect가 없다는 게 명확하다면 없다는 것을 명시해주는 것이 tree shaking에 도움이 된다. 

sideEffects를 package.json에서 false로 설정하면 package와 dependency들이 side effect를 발생하지 않는다는 것이다. (모든 모듈이 사이드 이펙트를 발생시키지 않는다.) 이렇게 표시해주면 번들러가 모든 모듈에 side effect가 없다고 판단하여 적절히 tree shaking을 해준다. 

```jsx
{
  "name": "webpack-tree-shaking-example",
  "version": "1.0.0",
  "sideEffects": [
    "./src/utils/utils.js"
  ] // side effect가 있는 파일들
}
```

side effect가 있는 파일들은 package에서 지정해줄 수 있다. 

프로젝트에서 css 파일을 가져오기 위해 css-loader같은 것을 사용할 수 있다. 이때 side effect 목록에 css파일을 추가해줘야 프로덕션 모드에서 실수로 삭제되는 것을 방지할 수 있다. 

## tree shaking을 위한 여러 조건들.

1. 바벨 설정을 통해 ES6문법이 commonJS로 변환되지 않도록 막기 

commonJS의 require는 export하는 모든 모듈을 가져오게 한다. 이를 방지 하기 위해 바벨 설정을 해주자. 

예시로 lodash가 있는데 lodash같은 경우 commonJS로 내보내기를 하고 있지 않아 tree shaking이 되지 않는다. 이를 해결하기 위해 두가지 방법이 있다. 첫째는 지원되는 모듈(lodash-es)을 사용하면 된다. 두번째는 플러그인을 추가로 사용해주면 된다. (babel-plugin-lodash)

```jsx
// babel-plugin-lodash 
// 1️⃣ => 2️⃣로 변경해준다. 

1️⃣import { cloneDeep } from 'lodash'
2️⃣import cloneDeep from 'lodash/cloneDeep'

```

2. side effect를 발생시키는 요소가 있는지 확인

3. 필요한 모듈만 import 해서 가져오기 

이는 tree shaking 조건을 맞추기 위해서 인데 필요한 모듈 일부만 import 할경우 webpack이 import 되지 않은 모듈들은 사용되지 않는다고 판단하여 shaking된다. 

## 참고

- [https://webpack.kr/guides/tree-shaking/](https://webpack.kr/guides/tree-shaking/)
- [https://ui.toast.com/weekly-pick/ko_20180716](https://ui.toast.com/weekly-pick/ko_20180716)
- [https://ko.javascript.info/modules-dynamic-imports](https://ko.javascript.info/modules-dynamic-imports)
- [https://helloinyong.tistory.com/305](https://helloinyong.tistory.com/305)
- [https://ykwan0714.github.io/lodash-크기-줄이기/](https://ykwan0714.github.io/lodash-%ED%81%AC%EA%B8%B0-%EC%A4%84%EC%9D%B4%EA%B8%B0/)- [https://huns.me/development/2265](https://huns.me/development/2265)

**Comment by airman5573** [comment]
import를 했는데도 tree shaking이 되는 경우는 어떤 경우인가요?

**Comment by airman5573** [comment]
tree shaking에서 side effect란?

**Comment by airman5573** [comment]
development모드에서 tree shaking이 발생 하나요?

**Comment by airman5573** [comment]
웹팩이 필요한 이유는?

**Comment by airman5573** [comment]
HTTP2가 나왔음에도 Webpack이 필요한 이유는 무엇인가요?
